### PR TITLE
Improve performance of UnmanagedMemoryStream

### LINKED ISSUE #93624: Interlocked.Read from 32-bit apps are much slower on Intel Sapphire Rapids CPUs
### Description

Interlocked.Read from 32-bit apps are much slower (100x) on Intel Sapphire Rapids CPUs.

```cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace UnmanagedMemoryStreamPerfTest
{
    class Program
    {
        static bool _Is32BitProcess = true;
        static Stream _Stream = null;
        static void Main(string[] args)
        {
            _Is32BitProcess = GetIs32BitProcess();
            _Stream = GetStream();

            Console.WriteLine("Is 32 bit process: " + _Is32BitProcess);
            Stopwatch sw = new Stopwatch();
            sw.Start();

            MainAsync();

            sw.Stop();
            Console.WriteLine(sw.ElapsedMilliseconds);
            Console.ReadLine();

        }

        static async void MainAsync()
        {
            //100 threads to make it eaven more obvious
            List<Task> tasks = new List<Task>();
            for (int i = 0; i < 100; i++)
                tasks.Add(DoStuff(GetStream()));
            await Task.WhenAll(tasks);
        }


        static unsafe async Task DoStuff(Stream stream)
        {
            //contrived, but this is just so we can see it.
            //for every XAML file, WPF will read baml streams from the assembly.
            //for apps with dense UIs, that can thousands of streams to open a screen

             //the problem, 32-bit will be about 3-20 times slower.  200 times slower on high end 2023 AMD and Intel chips due to a throttling feature for the type of lock UnmangedMemoryStream takes.
            int count = 1000000;
            for (int i = 0; i < count; i++)
            {
                DoStuffStream(stream);
                //DoStuffInterlocked()
                //DoStuffInterlockedTreF()
            }

        }

        static unsafe long DoStuffStream(Stream stream)
        {
            return stream.Length;
        }

        static unsafe long DoStuffInterlocked()
        {
            long len = 0;
            Interlocked.Read(ref len);
            return len;
        }

        static unsafe long DoStuffInterlockedTreF()
        {
            // inspriation from Parallel.For() fix  //https://devdiv.visualstudio.com/DevDiv/_search?text=969699&type=workitem&pageSize=25&filters=Projects%7BDevDiv%7D
            //Read uses Add under the hood, so this should be ok

            //gotcha is that 32-bit apps could then only use 2GB streams, limited by 32-bit int
            //maybe another Flag _BigStream

            long len = 0;
            if (_Is32BitProcess  /*&& !_BigStream*/)
            {
                long* indexPtr = &len;
                {
                    Interlocked.Add(ref *(int*)indexPtr, 0);
                }
            }
            else
            {
                Interlocked.Read(ref len);
            }
            return len;
        }

        static unsafe Stream GetStream()
        {
            //https://learn.microsoft.com/en-us/dotnet/api/system.io.unmanagedmemorystream?view=net-7.0
            //this is over kill, but its copy/paste MS sample code...

            byte[] message = UnicodeEncoding.Unicode.GetBytes(GetMessage());

            // Allocate a block of unmanaged memory and return an IntPtr object.	
            IntPtr memIntPtr = Marshal.AllocHGlobal(message.Length);

            // Get a byte pointer from the IntPtr object.
            byte* memBytePtr = (byte*)memIntPtr.ToPointer();

            // Create an UnmanagedMemoryStream object using a pointer to unmanaged memory.
            UnmanagedMemoryStream writeStream = new UnmanagedMemoryStream(memBytePtr, message.Length, message.Length, FileAccess.Write);

            // Write the data.
            writeStream.Write(message, 0, message.Length);

            // Close the stream.
            writeStream.Close();
            writeStream.Dispose();

            // Create another UnmanagedMemoryStream object using a pointer to unmanaged memory.
            UnmanagedMemoryStream readStream = new UnmanagedMemoryStream(memBytePtr, message.Length, message.Length, FileAccess.Read);

            // Create a byte array to hold data from unmanaged memory.
            byte[] outMessage = new byte[message.Length];

            // Read from unmanaged memory to the byte array.
            readStream.Read(outMessage, 0, message.Length);

            // Close the stream.
            //readStream.Close();
            //readStream.Dispose();

            return readStream;
        }

        static string GetMessage()
        {
            return "Here is some data.";
        }


        static bool GetIs32BitProcess()
        {
            return IntPtr.Size == 4;
        }
    }
}
```

### Configuration
.NET 4.8 and 6.0

### Regression?
No, new CPU behavior.

### Data
WPR of real world apps are available.

### Analysis

This was found tuning WPF applications on VDI and seeing a noticeable performance drop opening new screens on these new processors.
Interlocked.Read is used by a lot of APIs and apps.  If we can improve perf here, a lot of apps will magically run faster on modern hardware.  Otherwise, this issue will have to be addressed on an individual basis.

Many large enterprise apps have 32-bit 3rd-party dependencies, so cannot be rebuilt as 64-bit processes to avoid this perf issue.

### PR DESCRIPTION
**Pull request by jkotas**
UnmanagedMemoryStream used Interlocked operations to update its state to prevent tearing of 64-bit values on 32-bit platforms. This pattern is expensive in general and it was found to be prohibitively expensive on recent hardware..

This change removes the expensive Interlocked operations and addresses the tearing issues in alternative way:
- The _length field is converted to nuint that is guaranteed to be updated atomically.
- Writes to _length field are volatile to guaranteed the unininitialized memory cannot be read.
- The _position field remains long and it has a risk of tearing. It is not a problem since tearing of this field cannot lead to buffer overruns.

Fixes #93624

### CODE REVIEW

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/IO/UnmanagedMemoryStream.cs [hunk 1]
```diff
@@ -367,8 +361,8 @@ internal int ReadCore(Span<byte> buffer)
 
             // Use a local variable to avoid a race where another thread
             // changes our position after we decide we can read some bytes.
-            long pos = Interlocked.Read(ref _position);
-            long len = Interlocked.Read(ref _length);
+            long pos = _position;
+            long len = (long)Volatile.Read(ref _length);
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/IO/UnmanagedMemoryStream.cs:365]: Reordering of `NativeMemory.Clear` and `_length`. It is counterpart of `Volative.Write` used to set the `_length`.

If this was a regular read, I think it is theoretically possible for the code below to see uninitialized buffer.

**Comment by gautham-beeraka** [comment]
Thank you for the fix. Will the fix be part of .NET8? Will it also be ported to .NET Framework?

**Comment by jkotas** [comment]
> Will the fix be part of .NET8?

I will do a bar-check about backporting this fix to .NET 8.

> Will it also be ported to .NET Framework?

I do not expect this fix to be ported to .NET Framework. Performance fixes are ported to .NET Framework only in very exceptional situations.

File: src/libraries/System.Private.CoreLib/src/System/IO/UnmanagedMemoryStream.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/IO/UnmanagedMemoryStream.cs:369]: Comment added

**Review by jeffhandley — APPROVED** [review]

**Review by jozkee — APPROVED** [review]

**Comment by jkotas** [comment]
> I will do a bar-check about backporting this fix to .NET 8.

The fix was approved for .NET 8. https://github.com/dotnet/runtime/pull/93812

**Pull request merged**